lib/labelhash.cc

```diff
@@ -336,3 +346,176 @@ LabelHash::~LabelHash()
+        outfile.write(buf, n_bytes);
+        n_bytes = 0;
+      }
+    }
+    // save remainder.
+    if (n_bytes) {
+        outfile.write(buf, n_bytes);
+    }
+
+    if (outfile.fail()) {
+        delete[] buf;
+        throw khmer_file_exception(strerror(errno));
+    }
+    outfile.close();
+
+    delete[] buf;
+}
+
+void LabelHash::load_labels_and_tags(std::string filename)
+{
+    ifstream infile;
+
+    // configure ifstream to raise exceptions for everything.
+    infile.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+
+    try {
+        infile.open(filename.c_str(), ios::binary);
+    }  catch (std::ifstream::failure &e) {
+        std::string err;
+        if (!infile.is_open()) {
+            err = "Cannot open labels/tags file: " + filename;
+        } else {
+            err = "Unknown error in opening file: " + filename;
+        }
+        throw khmer_file_exception(err);
+    }
+
+    try {
+        unsigned int save_ksize = 0;
+        unsigned char version, ht_type;
+
+        infile.read((char *) &version, 1);
+        infile.read((char *) &ht_type, 1);
+        if (!(version == SAVED_FORMAT_VERSION)) {
+            std::ostringstream err;
+            err << "Incorrect file format version " << (int) version
+                << " while reading labels/tags from " << filename;
+            throw khmer_file_exception(err.str());
+        } else if (!(ht_type == SAVED_LABELSET)) {
+            std::ostringstream err;
+            err << "Incorrect file format type " << (int) ht_type
+                << " while reading labels/tags from " << filename;
+            throw khmer_file_exception(err.str());
+        }
+
+        infile.read((char *) &save_ksize, sizeof(save_ksize));
+        if (!(save_ksize == graph->ksize())) {
+            std::ostringstream err;
+            err << "Incorrect k-mer size " << save_ksize
+                << " while reading labels/tags from " << filename;
+            throw khmer_file_exception(err.str());
+        }
+    } catch (std::ifstream::failure &e) {
+        std::string err;
+        err = "Unknown error reading header info from: " + filename;
+        throw khmer_file_exception(err);
+    }
+
+    char * buf = new char[IO_BUF_SIZE];
+
+    unsigned int loaded = 0;
+    long remainder;
+
+
+    HashIntoType * kmer_p = NULL;
+    Label * labelp = NULL;
+
+    remainder = 0;
+    unsigned int iteration = 0;
+    while (!infile.eof()) {
+      //std::cout << "XXX " << loaded << " : " << "\n";
+        unsigned int i;
+
+        try {
+            infile.read(buf + remainder, IO_BUF_SIZE - remainder);
+        } catch (std::ifstream::failure &e) {
+
+            // We may get an exception here if we fail to read all the
+            // expected bytes due to EOF -- only pass it up if we read
+            // _nothing_.  Note that the while loop exits on EOF.
+
+            if (infile.gcount() == 0) {
+                std::string err;
```

Comment:
I'm happy to make them for you

On Wed, May 27, 2015 at 12:43 PM C. Titus Brown <notifications@github.com>
wrote:

> In lib/labelhash.cc
> <https://github.com/dib-lab/khmer/pull/1021#discussion_r31154113>:
>
> > +    remainder = 0;
> > +    unsigned int iteration = 0;
> > +    while (!infile.eof()) {
> > +      //std::cout << "XXX " << loaded << " : " << "\n";
> > +        unsigned int i;
> > +
> > +        try {
> > +            infile.read(buf + remainder, IO_BUF_SIZE - remainder);
> > +        } catch (std::ifstream::failure &e) {
> > +
> > +            // We may get an exception here if we fail to read all the
> > +            // expected bytes due to EOF -- only pass it up if we read
> > +            // _nothing_.  Note that the while loop exits on EOF.
> > +
> > +            if (infile.gcount() == 0) {
> > +                std::string err;
>
> Again, don't want to have to gin up a broken input file here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1021/files#r31154113>.
>